### PR TITLE
E2E testing

### PR DESCRIPTION
**Pull request by J3fftw1**
## Description
We want to start putting more velocity into PRs, this means we also want more validation around the PRs we want to merge in.

Unit tests are good but can't catch everything, they can all pass but the plugin fails to start for XYZ reason. Therefore, we want to start adding some e2e tests for these cases where unit tests cannot provide the validation we want.

## Proposed changes
Adds a new workflow to test all supported MC versions (+ Java versions) in a live server environment.
This uses a new "e2e-tester" tool which is currently private within the org but will soon be open.

## Checklist
- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [x] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

**Comment by WalshyDev** [comment]
TODO list:
- [ ] Version matrix
- [ ] Add an initial test which just ensures plugin loads and `/sf versions` can be ran

**Comment by WalshyDev** [comment]
It's coming together
image

File: .github/workflows/e2e-testing.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+          mv e2e-tester.jar plugins/e2e-tester.jar
+
+      - uses: actions/download-artifact@v3
+        with:
+          name: slimefun-${{ github.event.number }}
+          path: plugins/
+
+      - name: Run server
+        run: |
+          java -jar paper.jar --nogui
```
Comment (J3fftw1) [inline .github/workflows/e2e-testing.yml:64]: reminder for later
Setup server first
Then download everything
Paper > Slimefun > Test Plugin
Then run the server
The other is now all over the place

File: .github/workflows/e2e-testing.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+          - mcVersion: '1.19.4'
+            javaVersion: '19'
+          - mcVersion: '1.20.1'
+            javaVersion: '20'
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Set up JDK 17
```
Comment (J3fftw1) [inline .github/workflows/e2e-testing.yml:36]:
```suggestion
      - name: Set up JDK
```

File: .github/workflows/e2e-testing.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  workflow_call:
+    inputs:
+      artifact-name:
+        description: 'Slimefun artifact name'
+        required: true
+        type: string
+
+jobs:
+  e2e-testing:
+
```
Comment (J3fftw1) [inline .github/workflows/e2e-testing.yml:13]:
```suggestion
```

File: .github/workflows/e2e-testing.yml [hunk 4]
```diff
@@ -0,0 +1,71 @@
+name: End to End Testing
```
Comment (J3fftw1) [inline .github/workflows/e2e-testing.yml:1]:
```suggestion
name: End to End Testings
```

File: .github/workflows/e2e-testing.yml [hunk 5]
```diff
@@ -0,0 +1,71 @@
+name: End to End Testings
```
Comment (J3fftw1) [inline .github/workflows/e2e-testing.yml:1]:
```suggestion
name: End to End Testing
```

File: .github/workflows/e2e-testing.yml [hunk 6]
```diff
@@ -0,0 +1,71 @@
+name: End to End Testing
+
+on:
+  workflow_call:
+    inputs:
+      artifact-name:
+        description: 'Slimefun artifact name'
```
Comment (J3fftw1) [inline .github/workflows/e2e-testing.yml:7]:
```suggestion
        description: 'Slimefun artifact names'
```

File: .github/workflows/e2e-testing.yml [hunk 7]
```diff
@@ -0,0 +1,71 @@
+name: End to End Testing
+
+on:
+  workflow_call:
+    inputs:
+      artifact-name:
+        description: 'Slimefun artifact names'
```
Comment (J3fftw1) [inline .github/workflows/e2e-testing.yml:7]:
```suggestion
        description: 'Slimefun artifact name'
```

**Review by JustAHuman-xD — APPROVED** [review]
If anything goes wrong blame Jeff KEKW

**Review by Sfiguz7 — APPROVED** [review]
Totally not at gunpoint

**Comment by JustAHuman-xD** [comment]
> Totally not at gunpoint

🤣

**Pull request merged**